### PR TITLE
Add IDLE support

### LINKED ISSUE #157: IDLE support
This is one of the more exciting features for me. I was looking into the PR that's currently open for it #136 and it got me thinking *how should this actually work?*

I was pretty unaware how IDLE works, but [this stack overflow answer](http://stackoverflow.com/a/21345164) explains the principles pretty succinctly. Also checkout [the source code for `idle` and `idle_done`](https://github.com/ruby/ruby/blob/trunk/lib/net/imap.rb#L918-L959) 

Basically it opens a blocking connection that listens for events in the mailbox (which is pretty obvious), but less obvious is that you can't reuse that connection until you stop idling. Meaning that even within the handler for a new message you cannot do something like a `uid_fetch` (at least not using that came connection). Also IDLE connections will close down after 30 minutes so if you want to have them always running, the best practice is to set a timeout for 29 minutes and close/reopen the IDLE connection.

There's some complexity here so my question is how much of this should we abstract away in this gem vs leave for the user to handle?

Some ideas (with pseudocode):
```ruby
# blocks until it receives the next message then shuts down the connection and
# either calls the block if one was given or returns the response
def idle(mailbox, &block)
  imap.idle do |response|
    # Message received!
    imap.idle_stop

    if block_given?
      yield response
    else
      response
    end
  end
end

# same as the above but loops forever
def idle(mailbox, &block)
  # need a block here because this method doesn't return
  raise LocalJumpError, "no block given" unless block
  loop do
    imap.idle do |response|
      # Message received!
      imap.idle_stop

      yield response
    end
  end
end

# same as the above but give the user an option to loop forever or not
def idle(mailbox, options = {}, &block)
  # need a block here because this method doesn't return
  raise LocalJumpError, "no block given" unless block
  loop do
    imap.idle do |response|
      # Message received!
      imap.idle_stop

      yield response
    end

    break unless options[:keepalive]
  end
end
```

Additional thoughts:
- The `yield response` could potentially be wrapped inside a separate IMAP connection so that you can interact with message as you'd expect and we wouldn't have to call `idle_stop`.
- These can be thrown inside threads so they're non-blocking for the user. Or we can use EventMachine (in fact, someone is already doing that for this: https://github.com/ConradIrwin/em-imap).

@gmailgem/owners I'd love to get your thoughts here :thought_balloon:

### PR DESCRIPTION
**Pull request by alexeymorozov**
This PR resolves #157. Look at the discussion there.

@gmailgem/owners I would like to listen to your feedback on it :sunny:

### CODE REVIEW

**Comment by johnnyshields** [comment]
@jcarbo any thoughts on this?

File: lib/gmail/mailbox.rb [hunk 1]
```diff
@@ -118,6 +118,46 @@ def expunge
...
+            end
+          elsif resp.kind_of?(Net::IMAP::UntaggedResponse) && resp.name == 'EXISTS'
+            response = resp
+
+            @gmail.conn.synchronize do
+              complete_now = true
+              complete_cond.signal
+            end
+          end
+        end
```
Comment (jcarbo) [inline lib/gmail/mailbox.rb]: @alexeymorozov - I haven't done much threaded/concurrent programming. Can you explain why you need to open a new `Thread` and use `new_cond`? i.e. why can't you just do something like:
```ruby
@gmail.conn.idle do |resp|
  if resp.kind_of?(Net::IMAP::ContinuationRequest) && resp.data.text == 'idling'
    next
  elsif resp.kind_of?(Net::IMAP::UntaggedResponse) && resp.name == 'EXISTS'
    response = resp
    @gmail.conn.idle_done
  end
end
```

Comment (alexeymorozov) [inline lib/gmail/mailbox.rb]: This is how I implemented the 29-minute re-issue. A new thread is waiting for 29 minutes or the signal and then stops the IDLE command. Did I clarify it? Ask me more please.

**Comment by johnnyshields** [comment]
I'm going to merge this for now as the code and tests look good. I'll be testing in staging/production the next few weeks. Thanks @alexeymorozov

**Comment by bootstraponline** [comment]
This has been released in version 0.6.0 of the gmail gem. Thanks for contributing!

**Pull request merged**